REGRESSION (iOS 16): Unable to search in PDFs in Safari

pxlcoder opened the pull request:
#### 19bc5f6241e523672c8b4fcda0fc7ed53ac8d5fb
<pre>
REGRESSION (iOS 16): Unable to search in PDFs in Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=243522">https://bugs.webkit.org/show_bug.cgi?id=243522</a>
rdar://98018313

Reviewed by Wenson Hsieh.

Safari adopted UIFindInteraction to support find-in-page in iOS 16. UIKit also
provides new UIResponderStandardEditActions in iOS 16 to support find actions.
Currently, WebKit only responds to these actions if there is a WKContentView,
as WKWebView forwards `canPerformAction:withSender:` to the content view.
However, if a PDF is loaded, there is no content view, and the find actions have
no effect. Previously, Safari overrode the actions to grab the find interaction
and present the find navigator. That logic was removed to take full advantage of
the new find API, but revealed a flaw in our find-in-PDF support.

To fix, move the find actions (with the exception of &quot;find using selection&quot;, which
is unsupported due to a lack of functionality in PDFHostViewController) from
WKContentView into WKWebView. This approach ensures that the find navigator is
presented even if the content is in a WKPDFView.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView canPerformAction:withSender:]):
(-[WKWebView find:]):
(-[WKWebView findNext:]):
(-[WKWebView findPrevious:]):
(-[WKWebView findAndReplace:]):
(-[WKWebView performTextSearchWithQueryString:usingOptions:resultAggregator:]):

Use the current searchable object here, so that the WKPDFView is used for
searching when a PDF is loaded.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):
(-[WKContentView _findSelectedForWebView:]):
(-[WKContentView findForWebView:]): Deleted.
(-[WKContentView findNextForWebView:]): Deleted.
(-[WKContentView findPreviousForWebView:]): Deleted.
(-[WKContentView findAndReplaceForWebView:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
(swizzledPerformTextSearchWithQueryString):
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPDFView.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/253114@main">https://commits.webkit.org/253114@main</a>
</pre>

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
```diff
@@ -807,4 +808,25 @@ static void testPerformTextSearchWithQueryStringInWebView(WKWebView *webView, NS
+    // (controller) cannot be created in TestWebKitAPI, and we cannot actually search the PDF.
+    std::unique_ptr<InstanceMethodSwizzler> isInBackgroundSwizzler = makeUnique<InstanceMethodSwizzler>(NSClassFromString(@"WKPDFView"), @selector(performTextSearchWithQueryString:usingOptions:resultAggregator:), reinterpret_cast<IMP>(swizzledPerformTextSearchWithQueryString));
+
+    RetainPtr<WKWebView> webView = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600)]);
```
Nit - `auto` here (or just `RetainPtr`, for consistency with `searchOptions` below)

pxlcoder replied:
Done. Made this change in the other test too, and used `auto` with `searchOptions` below.

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10514,7 +10488,7 @@ - (void)useSelectionForFindForWebView:(id)sender
 {
     [self useSelectionForFindForWebView:sender];
-    [self findForWebView:sender];
+    [self.webView find:nil];
```
Are we intentionally dropping the `sender` here?

pxlcoder replied:
Not intentional, I put it back.

pxlcoder committed:
REGRESSION (iOS 16): Unable to search in PDFs in Safari
https://bugs.webkit.org/show_bug.cgi?id=243522
rdar://98018313

Reviewed by Wenson Hsieh.

Safari adopted UIFindInteraction to support find-in-page in iOS 16. UIKit also
provides new UIResponderStandardEditActions in iOS 16 to support find actions.
Currently, WebKit only responds to these actions if there is a WKContentView,
as WKWebView forwards `canPerformAction:withSender:` to the content view.
However, if a PDF is loaded, there is no content view, and the find actions have
no effect. Previously, Safari overrode the actions to grab the find interaction
and present the find navigator. That logic was removed to take full advantage of
the new find API, but revealed a flaw in our find-in-PDF support.

To fix, move the find actions (with the exception of "find using selection", which
is unsupported due to a lack of functionality in PDFHostViewController) from
WKContentView into WKWebView. This approach ensures that the find navigator is
presented even if the content is in a WKPDFView.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView canPerformAction:withSender:]):
(-[WKWebView find:]):
(-[WKWebView findNext:]):
(-[WKWebView findPrevious:]):
(-[WKWebView findAndReplace:]):
(-[WKWebView performTextSearchWithQueryString:usingOptions:resultAggregator:]):

Use the current searchable object here, so that the WKPDFView is used for
searching when a PDF is loaded.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):
(-[WKContentView _findSelectedForWebView:]):
(-[WKContentView findForWebView:]): Deleted.
(-[WKContentView findNextForWebView:]): Deleted.
(-[WKContentView findPreviousForWebView:]): Deleted.
(-[WKContentView findAndReplaceForWebView:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
(swizzledPerformTextSearchWithQueryString):
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPDFView.mm:
(TEST):

Canonical link: https://commits.webkit.org/253114@main
